docs/content/sips/013-sqlite.md

```diff
@@ -0,0 +1,165 @@
+
+```fsharp
+// A handle to an open sqlite instance
+type connection = u32
+
+// The set of errors which may be raised by functions in this interface
+variant error {
+  // A database with the supplied name does not exist
+  no-such-database,
+  // The requesting component does not have access to the specified database (which may or may not exist).
+  access-denied,
+  // The provided connection is not valid
+  invalid-connection,
+  // The database has reached its capacity
+  database-full,
+  // Some implementation-specific error has occurred (e.g. I/O)
+  io(string)
+}
+
+// Open a connection to a named database instance.
+//
+// If `database` is "default", the default instance is opened.
+//
+// `error::no-such-database` will be raised if the `name` is not recognized.
+open: func(name: string) -> expected<connection, error>
+
+// Execute a statement
+execute: func(conn: connection, statement: string, parameters: list<value>) -> expected<unit, error>
+
+// Query data
+query: func(conn: connection, query: string, parameters: list<value>) -> expected<query-result, error>
+
+// Close the specified `connection`.
+close: func(conn: connection)
+
+// A result of a query
+record query-result {
+  // The names of the columns retrieved in the query
+  columns: list<string>,
+  // The row results each containing the values for all the columns for a given row
+  rows: list<row-result>,
+}
+
+// A set of values for each of the columns in a query-result
+record row-result {
+  values: list<value>
+}
+
+// The values used in statements/queries and returned in query results
+variant value {
+  integer(s64),
+  real(float64),
+  text(string),
+  blob(list<u8>),
+  null
+}
+```
+
+*Note: the pseudo-resource design was inspired by the interface of similar functions in [WASI preview 2](https://github.com/bytecodealliance/preview2-prototyping/blob/d56b8977a2b700432d1f7f84656d542f1d8854b0/wit/wasi.wit#L772-L794).*
+
+#### Interface open questions
+
+* `row-result` can be very large. Should we provide some paging mechanism or a different API that allows for reading subsets of the returned data?
+  * Crossing the wit boundary could potentially be expensive if the results are large enough. Giving the user control of how they read that data could be helpful.
+* Is there really a need for query *and* execute functions since at the end of the day, they are basically equivalent?
+
+#### Database migrations
+
+Database tables typically require some sort of configuration in the form of database migrations to get table schemas into the correct state. To begin with a command line option supplied to `spin up` will be available for running SQL statements on start up (i.e., `--sqlite-migration "CREATE TABLE users..."`). It will be up to the user to provide itempotent statements such that running them multiple times does not produce unexpected results.
```

Comment:
Couple of ill-formed thoughts on this:

* As a point of detail, I suggest `--sqlite` rather than specifically talking about migrations, unless the flag is specifically bound to some sort of migrations format or infrastructure.  It's a pretty general purpose command facility.
* Because SQL statements can be quite long and fiddly, it would be good for users not to have to type them interactively on the CLI.  So it might be worth, even on day one, having a syntax for running a statement (or statements?) from a file, e.g. `--sqlite @usertable.sql`.
* Longer term, could consider a way to connect the user's favourite migration to hook up to the private database so it could generate the required scripts (or execute them directly?).  Not sure what this would look like though.
* `--key-value` currently operates only on the default store.  Are we okay with `--sqlite` operating only on the default database?  I guess if they have non-default ones then those are BYO and they can easily use their favourite SQLite tools on them.

Reply:
1.) I've renamed the option to just `sqlite` 👍
2.) I think this would be a good idea. Is `@` the right syntax for indicating a file?
3 and 4.) I think 3 is related to number 4. If we give the user a way to run sqlite statements (perhaps through a file) on any of their databases, we won't need a different mechanism for the non-default databases.

Reply:
Big +1 on something like `--sqlite @usertable.sql`.

Reply:
Done!! You can now run sql from files:

```
spin up --sqlite="@migration.sql"
```